compiler/noirc_evaluator/src/ssa/opt/pure.rs

```diff
@@ -0,0 +1,269 @@
+        // up a call graph of sorts to check afterwards to propagate impurity
+        // from called functions to their callers. Resultingly, an initial "Pure"
+        // result here could be overridden by one of these dependencies being impure.
+        let mut dependencies = BTreeSet::new();
+        let mut result = Purity::Pure;
+
+        for block in self.reachable_blocks() {
+            for instruction in self.dfg[block].instructions() {
+                // We don't defer to Instruction::can_be_deduplicated, Instruction::requires_acir_gen_predicate,
+                // etc. since we don't consider local mutations to be impure. Local mutations should
+                // be invisible to calling functions so as long as no references are taken as
+                // parameters or returned, we can ignore them.
+                // We even ignore Constrain instructions. As long as the external parameters are
+                // identical, we should be constraining the same values anyway.
+                match &self.dfg[*instruction] {
+                    Instruction::Constrain(..)
+                    | Instruction::ConstrainNotEqual(..)
+                    | Instruction::RangeCheck { .. } => {
+                        result = Purity::PureWithPredicate;
+                    }
+
+                    // These instructions may be pure unless:
+                    // - We may divide by zero
+                    // - The array index is out of bounds.
+                    // For both cases we can still treat them as pure if the arguments are known
+                    // constants.
+                    ins @ (Instruction::Binary(_)
+                    | Instruction::ArrayGet { .. }
+                    | Instruction::ArraySet { .. }) => {
+                        if ins.requires_acir_gen_predicate(&self.dfg) {
+                            result = Purity::PureWithPredicate;
+                        }
+                    }
+                    Instruction::Call { func, .. } => {
+                        match &self.dfg[*func] {
+                            Value::Function(function_id) => {
+                                // We don't know if this function is pure or not yet,
+                                // so track it as a dependency for now.
+                                dependencies.insert(*function_id);
+                            }
+                            Value::Intrinsic(intrinsic) => match intrinsic.purity() {
+                                Purity::Pure => (),
+                                Purity::PureWithPredicate => result = Purity::PureWithPredicate,
+                                Purity::Impure => return (Purity::Impure, BTreeSet::new()),
+                            },
+                            // The function we're calling is unknown in the remaining cases,
+                            // so just assume the worst.
+                            Value::ForeignFunction(_)
```

Comment:
Nit but I think it's worth separating out `ForeignFunction` to call out that foreign calls are 100% impure whereas the other cases are unreachable off of the top of my head.

Reply:
Done